odl/trafos/backends/pywt_bindings.py

```diff
@@ -0,0 +1,523 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Bindings to the ``PyWavelets`` backend for wavelet transforms."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+try:
+    import pywt
+    PYWAVELETS_AVAILABLE = True
+except ImportError:
+    PYWAVELETS_AVAILABLE = False
+
+
+__all__ = ('pywt_wbasis', 'pywt_coeff_shape_list',
+           'pywt_coeff_to_array', 'array_to_pywt_coeff',
+           'PAD_MODES_ODL2PYWT', 'PAD_MODES_ODL2PYWT',
+           'PYWT_SUPPORTED_PAD_MODES', 'PYWAVELETS_AVAILABLE')
+
+
+PAD_MODES_ODL2PYWT = {'constant': 'zpd',
+                      'periodic': 'ppd',
+                      'symmetric': 'sym',
+                      'order0': 'cpd',
+                      'order1': 'sp1',
+                      'pywt_periodic': 'per'}
+PAD_MODES_PYWT2ODL = {v: k for k, v in PAD_MODES_ODL2PYWT.items()}
+PYWT_SUPPORTED_PAD_MODES = PAD_MODES_ODL2PYWT.keys()
+
+
+def pywt_wbasis(wbasis):
+    """Convert ``wbasis`` a to `pywt.Wavelet` if it is not already one."""
+    if isinstance(wbasis, pywt.Wavelet):
+        return wbasis
+    else:
+        return pywt.Wavelet(wbasis)
+
+
+def pywt_max_level(*args, **kwargs):
+    """Return maximum number of scaling levels."""
+    if not PYWAVELETS_AVAILABLE:
```

Comment:
I think this is pretty much independent of implementation since in wavelet decomposition, one computes a downsampled-by-factor-2 convolution of input data with a filter, i.e. the convolution adds a little bit depending on the filter length, and then you halve the size. What one ends up with is `log2(input_len / (filter_len - 1))` as the "sensible" number of levels one can have. After that, the convolved data will have always the same size (something like `2 * (filter_len - 1)`), so there is no more size reduction. After that, one could, of course, just happily go on, but it doesn't make much sense.
Different implementations may or may not allow for this indefinite filtering, but the maximum number of reasonable levels is always the same.
Honestly, one could simply implement the function by hand and ditch the `pywt` helper:
```python
>>> import pywt
>>> for i in range(1000):
...     for f in range(1, 31):
...         if f <= 1 or i < f - 1:
...             n = 0
...         else:
...             n = int(np.log2(i / (f - 1)))
...         m = pywt.dwt_max_level(i, f)
...         if n != m:
...             print('mismatch for i={}, f={}'.format(i, f))

```
No printout.
The above basically replicates [this function](https://github.com/PyWavelets/pywt/blob/master/pywt/_extensions/c/common.c#L70-L75). Seems a bit overkill to implement it in C.